docs/intro-to-dune/index.md

```diff
@@ -0,0 +1,284 @@
+
+## `dune` files
+
+Besides `dune-project`, Dune also looks at the `dune` files in our project.
+Basically, `dune` files tell Dune about directories, executables, libraries,
+tests, and anything else of interest. For example, here's the `dune` file inside
+the root directory of your project:
+
+<<< @/../dune{clj}
+
+Like `dune-project`, a `dune` file consists of one or more *stanzas*. The only
+stanza here is
+[dirs](https://dune.readthedocs.io/en/stable/dune-files.html#dirs), which tells
+Dune which directories to include in the build. Note that the stanzas accepted
+in `dune` files are not the same as the ones accepted by `dune-project`. See all
+possible `dune` stanzas in Dune's [Stanza
+Reference](https://dune.readthedocs.io/en/stable/dune-files.html#dune).
+
+## `melange.emit` stanza
+
+Let's now take a look at the `src/dune` file:
+
+<<< @/../src/counter/dune{clj}
+
+The main stanza of interest for us is
+[melange.emit](https://dune.readthedocs.io/en/stable/melange.html#melange-emit),
+which tells Dune to turn our OCaml files into JavaScript files. The fields we
+give to `melange.emit` here are `target`, `libraries`, `preprocess`, and
+`module_systems`, which are ones that we need to use for pretty much every
+Melange project.
+
+A note about the `target` field: It is referenced in the `serve` npm script,
+whose command is
+
+```bash
+webpack serve --open --mode development --entry ./_build/default/src/output/src/Index.js
+```
+
+If you change the value of `target` to, say, `stuff`, the value of the `--entry`
+option must be changed to
+
+```
+./_build/default/src/stuff/src/Index.js
+```
+
+::: info
+In this dune file, we're only concerned with building JavaScript to run in the
+browser, but if we also wanted to build JavaScript to run on Node, we'd include
+another `melange.emit` stanza specifically for that. See
+[melange-opam-template](https://github.com/melange-re/melange-opam-template/blob/main/src/dune)
+for an example of building for Node.
+:::
+
+## Counter app directory
+
+Create a new directory in `src` called `counter`, which will contain a new app
+that only renders the [Counter component](/counter/). Then make sure this new
+directory contains:
+
+- `Counter.re` from `src`
+- `dune` file from `src`
+- `Index.re` to render the app to the DOM
+
+  <<< @/../src/counter/Index.re
+- `Makefile` to serve the app using webpack dev server
+
+  <<< @/../src/counter/Makefile{make}
+
+After you've added those files, your `src/counter` directory should look like
+this:
+
+```
+src/counter
+├─ Counter.re
+├─ dune
+├─ Index.re
+└─ Makefile
+```
+
+## Why Makefile
+
+You might be wondering why we're using a `Makefile` instead of adding another
+npm script. In the world of OCaml, usage of [the Make build automation
+tool](https://en.wikipedia.org/wiki/Make_(software)) is pretty much ubiquitous,
+and this is not likely to change. So let's take a short detour to explain the
+`Makefile` we just added.
```

Comment:
I would say that using Makefile is a personal preference, you can use anything really. 
 (justfile, even npm scripts if u want).

In fact, most people don't use Makefiles at all.

Reply:
The main reason why we add Makefiles in Melange projects is for the creation of the opam switch and initial installation of opam packages, because that's the only command we can't safely write as a Dune alias, as the `dune` exe might not be available yet.

Otherwise, all other commands in the makefile are just pointing to dune commands:

https://github.com/melange-re/melange-for-react-devs/blob/f7ac1a701e0c98d215dc4ce6c1031b6ac77a01e0/Makefile#L32-L58

Reply:
Instead of saying Make usage is ubiquitous, I just say it's common. https://github.com/melange-re/melange-for-react-devs/pull/6/commits/5ca4d90edbfb7b79c7ce36279256065829dab015